### ISSUE TITLE
Some confusion about the speed of DGCNN

### ISSUE THREAD

**Issue by JihaoLee** [pyg-team/pytorch_geometric#2172]
From now on, we recommend using our discussion forum (https://github.com/rusty1s/pytorch_geometric/discussions) for general questions.

## ❓ Questions & Help

Thank you for the code. I have some confusions about the speed of DGCNN. I found that the speed is much slower than [this implementation](https://github.com/AnTao97/dgcnn.pytorch). I used the `batch` parameter as you said in this [issue](https://github.com/rusty1s/pytorch_geometric/issues/1829). However, the speed is still very slow (about 6s a batch in inference). Besides, to my surprise, the speed of training (about 2s a batch) is faster than inference. Could you know how to slove it?

**Comment by rusty1s** [comment]
Can you show me a minimal example to reproduce?

**Comment by JihaoLee** [comment]
Thank you for your timely reply. I think that the speed is greatly affected by `batch_size`. This is an example code (on a single RTX 2080Ti). I used a small batch size in inference stage, so the speed is slower than training.
And I still have some confusions. I would like you to help me. Thanks in advance.
1. Why the speed is nearly equal between inference and training state when batch size is set to the same number?
2. Does your code support convolution? I find that some point cloud models use `nn.Conv1d` or `nn.Conv2d` to serve as `MLP`. I think that convolution may be more faster.
3. Why is the speed of DGCNN is slower than [this implementation](https://github.com/AnTao97/dgcnn.pytorch) (about 0.002s a sample in a batch in training stage)?
```
import os
import time
import torch
import argparse
import numpy as np
import torch.nn as nn
import torch.nn.functional as F

from torch.optim import Adam
from torch.utils.data import Dataset
from torch_geometric.data import Data, Batch, DataLoader
from torch_geometric.nn import DynamicEdgeConv, global_max_pool
from torch.nn import Sequential as Seq, Linear as Lin, ReLU, Dropout, BatchNorm1d as BN


os.environ['CUDA_VISIBLE_DEVICES'] = '0'
device = torch.device('cuda: 0' if torch.cuda.is_available() else 'cpu')


class SampleDataset(Dataset):
    def __init__(self):
        pass

    def __len__(self):
        return 80

    def __getitem__(self, idx):
        pos = torch.rand([4096, 7])
        y = torch.tensor(np.random.randint(0, args.num_classes, (4096, 1)))

        return Data(pos=pos.view(-1, 7), y=y.view(-1))


def MLP(channels, batch_norm=True):
    return Seq(*[
        Seq(Lin(channels[i - 1], channels[i]), ReLU(), BN(channels[i]))
        for i in range(1, len(channels))
    ])


class Net(torch.nn.Module):
    def __init__(self, out_channels, k=30, aggr='max'):
        super(Net, self).__init__()

        self.conv1 = DynamicEdgeConv(MLP([2 * 7, 64, 64]), k, aggr)
        self.conv2 = DynamicEdgeConv(MLP([2 * 64, 64, 64]), k, aggr)
        self.conv3 = DynamicEdgeConv(MLP([2 * 64, 64, 64]), k, aggr)
        self.lin1 = MLP([3 * 64, 1024])

        self.mlp = Seq(MLP([1024, 256]), Dropout(0.5), MLP([256, 128]),
                       Dropout(0.5), Lin(128, out_channels))

    def forward(self, data):
        pos, batch = data.pos, data.batch
        x0 = pos
        x1 = self.conv1(x0, batch)
        x2 = self.conv2(x1, batch)
        x3 = self.conv3(x2, batch)
        out = self.lin1(torch.cat([x1, x2, x3], dim=1))
        out = self.mlp(out)
        return F.log_softmax(out, dim=1)


def args_func():
    parser = argparse.ArgumentParser()
    parser.add_argument('--epochs', type=int, default=200)
    parser.add_argument('--train_batch_size', type=int, default=8)
    # parser.add_argument('--test_batch_size', type=int, default=1)       # speed 1.57s per batch
    # parser.add_argument('--test_batch_size', type=int, default=2)       # speed 0.80s per batch
    # parser.add_argument('--test_batch_size', type=int, default=4)       # speed 0.41s per batch
    # parser.add_argument('--test_batch_size', type=int, default=8)       # speed 0.21s per batch
    parser.add_argument('--test_batch_size', type=int, default=16)        # speed 0.12s per batch
    parser.add_argument('--num_classes', type=int, default=16)
    parser.add_argument('--lr', type=float, default=0.001)
    parser.add_argument('--lr_decay_factor', type=float, default=0.5)
    parser.add_argument('--lr_decay_step_size', type=int, default=50)
    parser.add_argument('--weight_decay', type=float, default=0)
    args = parser.parse_args()
    return args


def main(args):
    train_set = SampleDataset()
    train_loader = DataLoader(train_set, batch_size=args.train_batch_size, pin_memory=True, num_workers=4, shuffle=True)

    test_set = SampleDataset()
    test_loader = DataLoader(test_set, batch_size=args.test_batch_size, pin_memory=True, num_workers=4, shuffle=False)

    model = Net(args.num_classes, k=20)
    model = model.to(device)
    optimizer = Adam(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)

    criterion = nn.CrossEntropyLoss()
    criterion = criterion.to(device)

    for epoch in range(1, args.epochs + 1):
        if torch.cuda.is_available():
            torch.cuda.synchronize()

        t_start = time.perf_counter()
        print("training")
        train(args, model, optimizer, train_loader, criterion)
        print("infer")
        infer(args, model, test_loader, criterion)

        if torch.cuda.is_available():
            torch.cuda.synchronize()
        t_end = time.perf_counter()

        print('Epoch: {:03d}, Duration: {:.2f}'.format(epoch, t_end - t_start))

        if epoch % args.lr_decay_step_size == 0:
            for param_group in optimizer.param_groups:
                param_group['lr'] = args.lr_decay_factor * param_group['lr']


def train(args, model, optimizer, train_loader, criterion):
    model.train()
    for step, batch in enumerate(train_loader):
        batch = batch.to(device)
        optimizer.zero_grad()
        tic = time.time()
        out = model(batch)
        toc = time.time()
        loss = criterion(out, batch.y)
        loss.backward()
        optimizer.step()
        print("Speed: ", (toc - tic) / args.train_batch_size)


def infer(args, model, test_loader, criterion):
    model.eval()
    for step, batch in enumerate(test_loader):
        batch = batch.to(device)
        with torch.no_grad():
            tic = time.time()
            out = model(batch)
            toc = time.time()
        loss = criterion(out, batch.y)
        print("Speed: ", (toc - tic) / args.test_batch_size)


if __name__ == "__main__":
    args = args_func()
    main(args)
    print("Done!")
```

**Comment by rusty1s** [comment]
Thanks for providing the example code. Indeed, the major bottleneck of `DGCNN` is the dynamic edge creation via `knn` after each layer. It currently amounts to about 99% of the runtime of your script. We have plans to support a faster `knn` implementation in the future, but for now, there is nothing I can do about it.

Your linked repository computes the `knn` graph via pairwise-distance computation followed by a `topk` filtering. This approach can be, in fact, very fast, since it will trade runtime with memory consumption. In fact, the code will compute a potentially huge pair-wise distance matrix of shape `[num_nodes, num_nodes]`. Your linked repository has further differences: It expects examples to have the same number of points, which leads to faster runtime, but cannot be applied to examples of varying size (which are likely to happen in real-world scenarios).

To answer your questions:
1. They are in fact equal since the bottleneck is the `knn computation` rather than neural network execution.
2. 1D-Convolutions can't be faster than simple matrix multiplication, since they will default to matrix multiplication under the hood
3. See my previous comment.

If you want to speed up your model, I suggest to utilize the `knn` implementation in the [linked repository](https://github.com/AnTao97/dgcnn.pytorch/blob/master/model.py#L27-L33), and utilize `EdgeConv` instead of `DynamicEdgeConv`. Keep in mind that you need to reshape your feature matrix to `[batch_size, num_nodes, num_features]` though.

**Comment by JihaoLee** [comment]
Thank you very much for your detailed answer! I will have a try!

**Comment by JihaoLee** [comment]
Hi, sorry to bother you again.

When I tested the DGCNN model, I encountered overfitting phenomenon. Have you encountered a similar problem? The mF1 metric in training set can reach nearly 90%. However, this metric is only about 20% in test set. This phenomenon will occur regardless of whether I use the original `knn` or other `knn` implementations.  And 'weight decay' and 'dropout' also do not work. Do you know how to solve it?

**Comment by rusty1s** [comment]
Which dataset are you using? IMO, this sounds more like a problem of a different processing of training and test sets (potentially induced by Batch Norm).

**Comment by JihaoLee** [comment]
I use the ISPRS3D dataset. I tried to remove part of BN and all of BN layers. However, the overfitting phenomenon still occurs.

**Comment by zimonitrome** [comment]
Is there any praxis for using `num_workers` in `DynamicEdgeConv`?

It states that `batch` must be `None`: Does this mean a batch size of 0?

It states that input can not lie on the GPU: Does anyone know if it is faster to move the input to CPU, do the kNN calculation, move it back to GPU, compared to using num_workers=1?

**Comment by rusty1s** [comment]
A `batch` of `None` refers to `batch_size==1`.

It can be more computationally efficient to out-source the `kNN` calculation to the CPU (dependent on your `batch_size` and `num_nodes`). This is just something I would test in practice via runtime measurement. We are also working on a better `num_workers` interface for batched computation.
